### ISSUE TITLE
MissingRequiredOptionError when specifying required Value after enumerable Options

### ISSUE THREAD

**Issue by ericnewton76** [commandlineparser/commandline#91]
<a href="https://github.com/SuperFXMaster"><img src="https://avatars0.githubusercontent.com/u/6494919?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [SuperFXMaster](https://github.com/SuperFXMaster)**
_Tuesday Oct 11, 2016 at 19:24 GMT_
_Originally opened as https://github.com/gsscoder/commandline/issues/361_

----

The following options class:

``` C#
public class CommandLineOptions
{
    [Value(0, Required = true)]
    public string InputFileName { get; set; }

    [Option('o', "output")]
    public string OutputFileName { get; set; }

    [Option('i', "include", Separator = ',')]
    public IEnumerable<string> Included { get; set; }

    [Option('e', "exclude", Separator = ',')]
    public IEnumerable<string> Excluded { get; set; }
}
```

When provided to the default parser, with the commandline `app.exe --exclude=a,b InputFile.txt` results in a `MissingRequiredOptionError: A required value not bound to option name is missing`. I expected the library to accept `InputFileName = InputFile.txt` and `--exclude=a,b` as `Excluded = { "a", "b" }`. Is there something else required to make this work as I expect, or is it a bug?

**Comment by ericnewton76** [comment]
<a href="https://github.com/nemec"><img src="https://avatars3.githubusercontent.com/u/136378?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nemec](https://github.com/nemec)**
_Tuesday Oct 11, 2016 at 19:57 GMT_

----

Yes, I think that's a bug. If you set `Required=false` you'll see that a, b, and InputFile.txt are all part of Included. Additionally, the "EnableDashDash" setting doesn't work either:

``` csharp
void Main()
{
    var p = new Parser(s =>
    {
        s.EnableDashDash = true;
    });
    var result = p.ParseArguments<CommandLineOptions>("--exclude=a,b -- InputFile.txt".Split());
    result.WithNotParsed(e => e.Dump());
    result.WithParsed(d => d.Dump());
}

public class CommandLineOptions
{
    [Value(0, Required = true)]
    public string InputFileName { get; set; }

    [Option('o', "output")]
    public string OutputFileName { get; set; }

    [Option('i', "include", Separator = ',')]
    public IEnumerable<string> Included { get; set; }

    [Option('e', "exclude", Separator = ',')]
    public IEnumerable<string> Excluded { get; set; }
}
```

**Comment by ericnewton76** [comment]
<a href="https://github.com/SuperFXMaster"><img src="https://avatars0.githubusercontent.com/u/6494919?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [SuperFXMaster](https://github.com/SuperFXMaster)**
_Wednesday Oct 12, 2016 at 16:20 GMT_

----

As a workaround, the commandline `--exclude=a,b --output=OutFile InFile` works fine, just as `-e a b c -o OutFile InFile` does. It seems like the problem arises when the `--exclude` option is directly before the `Required` parameter.

**Comment by ericnewton76** [comment]
<a href="https://github.com/anthonylangsworth"><img src="https://avatars2.githubusercontent.com/u/1168640?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [anthonylangsworth](https://github.com/anthonylangsworth)**
_Saturday Oct 22, 2016 at 11:50 GMT_

----

This appears to be fixed in the master branch, @SuperFXMaster and @nemec . I cannot reproduce the issue. Please let me know if I am incorrect.

**Comment by ericnewton76** [comment]
<a href="https://github.com/nemec"><img src="https://avatars3.githubusercontent.com/u/136378?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nemec](https://github.com/nemec)**
_Saturday Nov 05, 2016 at 07:01 GMT_

----

I'm still having issues on the master branch  with the code snippet I posted above, as well as @SuperFXMaster's.

**Comment by ericnewton76** [comment]
<a href="https://github.com/hugener"><img src="https://avatars2.githubusercontent.com/u/5023998?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [hugener](https://github.com/hugener)**
_Monday Mar 13, 2017 at 18:21 GMT_

----

Had quick look into this and could let the example above to run by changing the order of parameters in Concat in the following line:
https://github.com/gsscoder/commandline/blob/master/src/CommandLine/Core/Tokenizer.cs#L54
The changed line becomes:
`return tokenizerResult.Map(tokens => values.Concat(tokens));`

I'm not sure it's valid for other cases though. Maybe someone who knows more about it can have a look. It would be nice to get a stable version.

**Comment by rmunn** [comment]
This will be fixed by #684 once it's merged in.
